tawazi/_dag/helpers.py

```diff
@@ -0,0 +1,298 @@
+        call_id (str): A unique identifier for the execution.
+            It can be used to distinguish the id of the call inside the thread.
+            It might be useful to debug and to exchange information between the main thread and the sub-threads (per-node threads)
+
+    Returns:
+        node_dict: dictionary with keys the name of the function and value the result after the execution
+    """
+    # 0.1 deepcopy the node_dict in order to modify the results inside every node and make the dag reusable
+    #     modified_node_dict are used to modify the values inside the ExecNode corresponding
+    #     to the input arguments provided to the whole DAG (ArgExecNode)
+    xns_dict = modified_node_dict or copy_non_setup_xns(node_dict)
+
+    # 0.2 prune the graph from the ArgExecNodes so that they don't get executed in the ThreadPool
+    precomputed_xns_ids = [id_ for id_ in graph if xns_dict[id_].executed]
+    for id_ in precomputed_xns_ids:
+        graph.remove_node(id_)
+
+    # 0.3 create variables related to futures
+    futures: Dict[Identifier, "Future[Any]"] = {}
+    done: Set["Future[Any]"] = set()
+    running: Set["Future[Any]"] = set()
+    futures_launch_time: Dict[Identifier, float] = {}
+
+    # 0.4 get the candidates root nodes that can be executed
+    # runnable_nodes_ids will be empty if all root nodes are running
+    runnable_xns_ids = graph.root_nodes()
+
+    with ThreadPoolExecutor(max_workers=max_concurrency, thread_name_prefix=call_id) as executor:
+        while len(graph):
+            # Attempt to run **A SINGLE** root node.
+
+            # 6. block scheduler execution if no root node can be executed.
+            #    this can occur in two cases:
+            #       1. if maximum concurrency is reached
+            #       2. if no runnable node exists (i.e. all root nodes are being executed)
+            #    in both cases: block until a node finishes
+            #       => a new root node will be available
+            num_running_threads = get_num_running_threads(futures)
+            num_runnable_nodes_ids = len(runnable_xns_ids)
+            if num_running_threads == max_concurrency or num_runnable_nodes_ids == 0:
+                # must wait and not submit any workers before a worker ends
+                # (that might create a new more prioritized node) to be executed
+                logger.debug(f"Waiting for ExecNodes {running} to finish. Finished running {done}")
+
+                # next_timeout is used to determine the timeout of the next node to be executed
+                next_timeout = get_next_timeout(futures_launch_time, futures, xns_dict)
+
+                done_, running = wait(running, return_when=FIRST_COMPLETED, timeout=next_timeout)
+
+                raise_timeout_error_conditionally(done_, running, next_timeout)
+
+                done = done.union(done_)
+
+            # TODO: Optimize this part! I can just check for a the node or nodes in done_!
```

Comment:
could you rewrite thius TODO ? I'm not sure I understand I understand the optimization possibilities in the first part

Reply:
done